content/en/blog/_posts/2023-04-11-kmsv2-beta.md

```diff
@@ -0,0 +1,187 @@
+---
+layout: blog
+title: "Kubernetes 1.27: KMS V2 Moves to Beta"
+date: 2023-04-11
+slug: kms-v2-moves-to-beta
+---
+
+**Authors:** Anish Ramasekar, Mo Khan, and Rita Zhang (Microsoft)
+
+With Kubernetes 1.27, we are moving Key Management Service (KMS) v2 API to beta.
+
+## What is KMS?
+One of the first things to consider when securing a Kubernetes cluster is encrypting etcd data at rest.
+KMS provides an interface for a provider to utilize a key stored in an external key service to perform this encryption.
+
+KMS v1 has been a feature of Kubernetes since version 1.10,
+and is currently in beta as of version v1.12.
+KMS v2 was introduced as alpha in v1.25.
+
+{{% alert title="Note" color="primary" %}}
+The KMS v2 API and implementation changed in incompatible ways in-between the alpha release in v1.25 and the beta release in v1.27. The design of KMS v2 has changed since [the previous blog post](https://kubernetes.io/blog/2022/09/09/kms-v2-improvements/) was written and it is not compatible with the design in this blog post. Attempting to upgrade from old versions with the alpha feature enabled will result in data loss.
+{{% /alert %}}
+
+## What’s new in `v2beta1`?
+The KMS encryption provider uses an envelope encryption scheme to encrypt data in etcd. The data is encrypted using a data encryption key (DEK). The DEKs are encrypted with a key encryption key (KEK) that is stored and managed in a remote KMS. With KMS v1, a new DEK is generated for each encryption. With KMS v2, a new DEK is only generated on server startup and when the KMS plugin informs the API server that a KEK rotation has occurred.
+
+{{% alert title="Caution" color="warning" %}}
+If you are running virtual machine (VM) based nodes that leverage VM state store with this feature, you must not use KMS v2.
+
+With KMS v2, the API server uses AES-GCM with a 12 byte nonce (8 byte atomic counter and 4 bytes random data) for encryption. The following issues could occur if the VM is saved and restored:
+
+1. The counter value may be lost or corrupted if the VM is saved in an inconsistent state or restored improperly. This can lead to a situation where the same counter value is used twice, resulting in the same nonce being used for two different messages.
+2. If the VM is restored to a previous state, the counter value may be set back to its previous value, resulting in the same nonce being used again.
+Although both of these cases are partially mitigated by the 4 byte random nonce, this can compromise the security of the encryption.
```

Comment:
```suggestion

Although both of these cases are partially mitigated by the 4 byte random nonce, this can compromise the security of the encryption.
```